internals/overlord/state/state.go

```diff
@@ -156,27 +158,40 @@ func (s *State) unlock() {
 }
 
 type marshalledState struct {
-	Data     map[string]*json.RawMessage `json:"data"`
-	Changes  map[string]*Change          `json:"changes"`
-	Tasks    map[string]*Task            `json:"tasks"`
-	Warnings []*Warning                  `json:"warnings,omitempty"`
-	Notices  []*Notice                   `json:"notices,omitempty"`
+	Data       map[string]*json.RawMessage    `json:"data"`
+	Changes    map[string]*Change             `json:"changes"`
+	Tasks      map[string]*Task               `json:"tasks"`
+	Warnings   []*Warning                     `json:"warnings,omitempty"`
+	Notices    []*Notice                      `json:"notices,omitempty"`
+	Identities map[string]*marshalledIdentity `json:"identities,omitempty"`
 
 	LastChangeId int `json:"last-change-id"`
 	LastTaskId   int `json:"last-task-id"`
 	LastLaneId   int `json:"last-lane-id"`
 	LastNoticeId int `json:"last-notice-id"`
 }
 
+// marshalledIdentity is used specifically for marshalling to the state
+// database file. Unlike apiIdentity, it should include secrets.
+type marshalledIdentity struct {
+	Access string                   `json:"access"`
+	Local  *marshalledLocalIdentity `json:"local,omitempty"`
+}
+
+type marshalledLocalIdentity struct {
+	UserID uint32 `json:"user-id"`
```

Comment:
I wonder if this should be a string. I know on linux it can only be a unsigned int, but go abstracts user id as a string (because of windows). Probably not a problem, but at least then we can distinguish between empty/default and root (0).

Reply:
I think you're right that we should be able to distinguish between empty and root (0). People probably should be able to specify a `superuser: {access: admin, local: {user-id: 0}}` user (at the moment I disallow that).

However, I think `*uint32` is better than than string. It'd be better if it were a number, not a string, in JSON/YAML, as it's the correct type (and less validation is required). We've also used `UserID *int` elsewhere quite a bit for this same thing, and it's worked out well (though it probably should be `*uint32` rather than `*int` for correctness on 32-bit systems).

As far as strings on Windows go, the whole ucred system depends on this being an integer (and Linux), so I don't think we need to open that can of worms.

In any case, I'll change this to a `*uint32` and a named identity with user-id 0.

Reply:
Also, this only applies to `apiLocalIdentity.UserID` (used by `Identity.UnmarshalJSON`), as that's where identity JSON is unmarshalled. Oh, and `client.LocalIdentity.UserID` in the subsequent PR, as that's where the YAML version is unmarshalled.